provider/tests/provider_test.go

```diff
@@ -0,0 +1,112 @@
+package tests
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/blang/semver"
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/integration"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	command "github.com/pulumi/pulumi-command/provider/pkg/provider"
+	"github.com/pulumi/pulumi-command/provider/pkg/version"
+)
+
+func provider() integration.Server {
+	v := semver.MustParse(version.Version)
+	return integration.NewServer(command.Name, v, command.NewProvider())
+}
+
+func urn(mod, res, name string) resource.URN {
+	m := tokens.ModuleName(mod)
+	r := tokens.TypeName(res)
+	if !tokens.IsQName(name) {
+		panic(fmt.Sprintf("invalid resource name: %q", name))
+	}
+	n := tokens.QName(name)
+	return resource.NewURN("test", "command", "",
+		tokens.NewTypeToken(
+			tokens.NewModuleToken(command.Name, m),
+			r),
+		n)
+}
+
+func TestLocalCommand(t *testing.T) {
+	cmd := provider()
+	urn := urn("local", "Command", "echo")
+	unknown := resource.NewOutputProperty(resource.Output{
+		Element: resource.NewStringProperty(""),
+		Known:   false,
+	})
+	c := resource.MakeComputed
+
+	create := func(preview bool, env resource.PropertyValue) p.CreateResponse {
+		resp, err := cmd.Create(p.CreateRequest{
+			Urn: urn,
+			Properties: resource.PropertyMap{
+				"create":      resource.NewStringProperty("echo hello, $NAME!"),
+				"environment": env,
+			},
+			Preview: preview,
+		})
+		require.NoError(t, err)
+		return resp
+	}
+
+	assert.Equal(t, resource.PropertyMap{
+		"create": resource.PropertyValue{V: "echo hello, $NAME!"},
+		"stderr": c(resource.PropertyValue{V: ""}),
+		"stdout": c(resource.PropertyValue{V: ""}),
+	},
+		create(true /* preview */, unknown).Properties)
+
+	fullProperties := resource.PropertyMap{
+		"create": resource.PropertyValue{V: "echo hello, $NAME!"},
+		"stderr": resource.PropertyValue{V: ""},
+		"stdout": resource.PropertyValue{V: "hello, world!"},
```

Comment:
A bit unfamiliar with this method of testing. Could you explain why `hello, world!` is stdout here? Perhaps some comments about what each major step is doing might be beneficial.

Reply:
`"stdout"` is `"hello, world!"` because the command provider runs `echo hello, $NAME!` with `NAME=world` set in the enclosing environment.

Reply:
There aren't really major steps here. Instead it's a series of logically (but not semantically) tests. I'll refactor to make that clearer.